# Clase 2 — Tipos de Datos, Operadores y Modelo de Objetos

Bienvenido a la segunda clase del curso de Python. En esta sesión exploraremos el modelo de objetos, los tipos básicos, los operadores y cómo manejar cadenas de texto en profundidad.

**Contenido de esta clase:**
1. Uso de `uv` con Jupyter Notebooks
2. Extensión de archivos y Punto de Entrada
3. La función `print()`
4. Tipado dinámico y Modelo de Objetos
5. Palabras Reservadas
6. Tipos de Datos Primitivos (en profundidad)
7. Conversión de Tipos
8. Operadores (Aritméticos, Comparación, Lógicos, Bits, Asignación)
9. Cadenas de Texto (Strings) en profundidad
10. Ejercicio Práctico: Calculadora de Calificaciones Universitarias

---
## 1. Uso de `uv` con Jupyter Notebooks

`uv` es un gestor de paquetes de Python extremadamente rápido escrito en Rust (como vimos en la clase 1). Para usar `uv` junto con Jupyter Notebooks y mantener nuestro entorno aislado, seguimos estos pasos:

1. **Inicializar un proyecto y entorno virtual**:
   
   Al descargar este repositorio se debe usar lo siguiente para que uv instale las dependencias.
   ```bash
   uv sync
   ```
   Si es que deseas usar una carpeta `mi_proyecto` y ahí podrás programar si lo deseas.
   ```bash
   uv init mi_proyecto
   cd mi_proyecto
   uv venv
   ```
   O si deseas usar la carpeta actual
   ```bash
   uv init
   ```

2. **Activar el entorno** (Opcional, `uv run` lo maneja automáticamente):
   - Windows: `.venv\Scripts\activate`
   - Linux/macOS: `source .venv/bin/activate`
3. **Instalar Jupyter y el kernel de IPython usando uv**:
   ```bash
   uv add jupyter ipykernel
   ```
4. **Registrar el kernel** (para que Jupyter reconozca el entorno de uv):
   Esto en el caso de que quieras usar un entorno externo, si utilizas VS Code, entonces sólo hace falta elegir el entorno en el propio jupyter notebook. Si usas VS Code usa extensiones especializadas.
   ```bash
   uv run python -m ipykernel install --user --name=mi_proyecto_uv
   ```
5. **Iniciar Jupyter Notebook**:
   Para correrlo en el navegador.
   ```bash
   uv run jupyter notebook
   ```

Una vez en Jupyter, asegúrate de seleccionar el kernel `mi_proyecto_uv` en la esquina superior derecha (navegador).

---
## 2. Extensión de archivos y Punto de Entrada

Los scripts en Python se guardan con la extensión **`.py`**. Son archivos de texto que son interpretados por el intérprete de Python (Corcuera, IntroPython).

Cuando ejecutas un archivo directamente (por ejemplo: `python script.py`), Python asigna un nombre especial a ese módulo: `"__main__"`.

Esto nos permite definir un **punto de entrada** explícito, similar a la función `main()` en C.

> **Puente con C:** En C, el punto de entrada es siempre la función `main()`. En Python, el código se ejecuta de arriba a abajo, pero la convención `if __name__ == '__main__':` nos permite controlar qué código se ejecuta solo cuando el archivo es el programa principal (y no cuando es importado como módulo).

In [1]:
# mi_modulo.py
def main():
    print("¡Este es el punto de entrada principal! Solo se ejecutará si este archivo se ejecuta directamente.")

print("¡Este código se ejecutará incluso si se importa como módulo!")

if __name__ == '__main__':
    main()

¡Este código se ejecutará incluso si se importa como módulo!
¡Este es el punto de entrada principal! Solo se ejecutará si este archivo se ejecuta directamente.


---
## 3. La función `print()`

Antes de sumergirnos en tipos y operadores, conozcamos la herramienta que usaremos para ver resultados: la función `print()`.

`print()` es una **función predefinida** (built-in) que envía texto a la salida estándar (la consola). Por defecto, a diferencia de C, python importa automáticamente la librería estándar. Sus características más comunes son:

| Uso | Ejemplo | Resultado |
|-----|---------|-----------|
| **Un solo valor** (agrega el salto de línea por defecto) | `print("Hola")` | `Hola` |
| **Varios valores** (separados por espacio) | `print("Hola", "Mundo")` | `Hola Mundo` |
| **Parámetro `sep`** (cambia el separador) | `print("a", "b", "c", sep="-")` | `a-b-c` |
| **Parámetro `end`** (cambia el final, por defecto `\n`) | `print("Hola", end=" ")` + `print("Mundo")` | `Hola Mundo` |
| **Formato con f-strings** | `print(f"2 + 2 = {2 + 2}")` | `2 + 2 = 4` |
| **Representación con `!r`** | `print(f"{'Hola'!r}")` | `'Hola'` |

> **Nota:** En una clase futura sobre funciones exploraremos `print()` a fondo y veremos todas las funciones predefinidas. Por ahora, con estos usos básicos tenemos suficiente para lo que viene.

Ejecuta la celda de abajo para ver `print()` en acción:

In [1]:
print("Hola")                          # Un solo valor
print("Hola", "Mundo")               # Varios valores
print("a", "b", "c", sep="-")      # Separador personalizado
print("Hola", end=" ")               # Sin salto de línea al final
print("Mundo")                        # Continúa en la misma línea
print(f"2 + 2 = {2 + 2}")            # f-string
print(f"{'Python'!r}")               # Representación con comillas

Hola
Hola Mundo
a-b-c
Hola Mundo
2 + 2 = 4
'Python'


---
## 4. Tipado dinámico y Modelo de Objetos

### Todo es un Objeto

En Python, **todo es un objeto** (incluso las funciones y los números). Cada objeto se caracteriza por tres propiedades fundamentales (Corcuera, IntroPython):

- **Identidad**: un identificador único en memoria (obtenible con `id()`).
- **Tipo**: la clase del objeto (obtenible con `type()`).
- **Valor**: el dato que almacena.

### Tipado Dinámico

Python tiene **tipado dinámico**, lo que significa que:
- El tipo está asociado al **valor (objeto)** en memoria, **no a la variable** en sí.
- No necesitas declarar el tipo de una variable antes de usarla.
- Una variable se crea automáticamente cuando un valor es asignado la primera vez.
- Una variable puede reasignarse a objetos de cualquier tipo.

> **Puente con C:** A diferencia de C, donde una variable reserva un espacio de memoria con un tamaño fijo para un tipo específico (`int x = 5;`), en Python **todas las variables actúan como punteros (referencias)** a objetos en memoria. No existen tipos primitivos "por valor". El recolector de basura (Garbage Collector) se encarga de liberar la memoria cuando un objeto ya no tiene referencias.

> **Concepto clave (MoureDev, Guía Junior 2026):** Python usa **tipado dinámico** (tipos verificados durante la ejecución) y **tipado fuerte** (conversiones de tipo deben ser explícitas). Esto contrasta con C que es de tipado estático (tipos verificados en compilación).

In [ ]:
# Ejemplo de tipado dinámico, los errores se detectan en tiempo de ejecución (runtime)
x = 1
print(f"x = {x}, \t\ttipo = {type(x)}") # \t es un tabulador para alinear la salida

x = 1.0
print(f"x = {x}, \ttipo = {type(x)}")

x = 'hello'
print(f"x = {x}, \ttipo = {type(x)}")

x = [1, 2, 3]
print(f"x = {x}, \ttipo = {type(x)}")

# Ejemplo de tipado fuerte, el complilador o el intérprete no permite operaciones entre tipos incompatibles
# x = 1 + '1'  # Esto generará un error de tipo (TypeError) en tiempo de ejecución


x = 1, 		tipo = <class 'int'>
x = 1.0, 	tipo = <class 'float'>
x = hello, 	tipo = <class 'str'>
x = [1, 2, 3], 	tipo = <class 'list'>


### Identidad de objetos: `id()` y el operador `is`

La función `id()` devuelve la identidad única de un objeto. El operador `is` compara si dos variables apuntan al **mismo objeto** en memoria (no solo si tienen el mismo valor).

In [3]:
a = [1, 2, 3]
b = [1, 2, 3]
c = a

print(f"a == b: {a == b}")
print(f"a is b: {a is b}")
print(f"a is c: {a is c}")
print(f"id(a): {id(a)}")
print(f"id(b): {id(b)}")
print(f"id(c): {id(c)}")

a == b: True
a is b: False
a is c: True
id(a): 1939765236288
id(b): 1939765236096
id(c): 1939765236288


### El operador asignación (`=`)

En Python **no es necesario declarar las variables** antes de usarlas. La asignación inicial crea la variable y enlaza el valor a la variable (Corcuera, IntroPython).

Python soporta varias formas de asignación:
- **Asignación simple:** `x = 5`
- **Asignación múltiple (por pares):** `b, c, d = 123, 4.5, 'Hello'`
- **Asignación en cadena:** `e = f = g = 123`
- **Eliminación:** `del x` elimina la variable

In [4]:
b, c, d = 123, 4.5, 'Hello'
print(f"b={b}, c={c}, d={d}")

e = f = g = 123
print(f"e={e}, f={f}, g={g}")

x = 8
print(f"x = {x}")
del x
# print(x)  # NameError: name 'x' is not defined

b=123, c=4.5, d=Hello
e=123, f=123, g=123
x = 8


---
## 5. Palabras Reservadas

Al igual que en C, existen palabras clave que no puedes usar como identificadores (nombres de variables o funciones). Las palabras reservadas en Python 3 son (Corcuera, IntroPython):

| | | | | | |
|---|---|---|---|---|---|
| `False` | `None` | `True` | `and` | `as` | `assert` |
| `async` | `await` | `break` | `class` | `continue` | `def` |
| `del` | `elif` | `else` | `except` | `finally` | `for` |
| `from` | `global` | `if` | `import` | `in` | `is` |
| `lambda` | `nonlocal` | `not` | `or` | `pass` | `raise` |
| `return` | `try` | `while` | `with` | `yield` | |

In [ ]:
import keyword
print(f"Total de palabras reservadas: {len(keyword.kwlist)}") #len() devuelve el número de elementos en la lista de palabras reservadas

# PEP 20: Flat is better than nested.
lista_primeras_10 = keyword.kwlist[:10] # Crea una lista con las primeras 10 palabras reservadas
lista_siguientes_10 = keyword.kwlist[10:20] # Crea una lista con las siguientes 10 palabras reservadas
lista_restantes = keyword.kwlist[20:30] # Crea una lista con las palabras reservadas restantes (si hay más de 20)

print(lista_primeras_10) # Imprime las primeras 10 palabras reservadas
print(lista_siguientes_10) # Imprime las palabras reservadas restantes
print(lista_restantes) # Imprime las palabras reservadas restantes

Total de palabras reservadas: 35
['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class']
['continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global']
['if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise']


---
## 6. Tipos de Datos Primitivos

Un **tipo de dato** es el conjunto de valores y el conjunto de operaciones definidas en esos valores (Corcuera, IntroPython). Python tiene un gran número de tipos de datos incorporados tales como Números (Integer, Float, Boolean, Complex Number), String, List, Tuple, Set, Dictionary. Otros tipos de datos de alto nivel, tales como Decimal y Fraction, están soportados por módulos externos.

Según la Guía de MoureDev (2026), los **tipos primitivos** fundamentales son:
- **Números enteros**: valores sin decimales, positivos o negativos.
- **Números decimales**: valores con punto flotante, precisión limitada.
- **Booleanos**: verdadero o falso, base de toda lógica.
- **Caracteres**: símbolos individuales de texto (en Python no existe un tipo `char` dedicado).
- **Nulo/Vacío**: ausencia intencional de valor.

### 6.1 `int` — Números Enteros

Tipo de dato (`int`) para representar enteros o números naturales (Corcuera, IntroPython).

**Características clave:**
- A diferencia de C (`int`, `long` con tamaños fijos de 32 o 64 bits), los enteros en Python 3 son de **tamaño ilimitado**.
- Se pueden expresar en diferentes bases:
  - **Decimal** (base 10): `123`
  - **Hexadecimal** con prefijo `0x` o `0X`: `0x1abc`, `0X1ABC`
  - **Octal** con prefijo `0o` o `0O`: `0o1776`
  - **Binario** con prefijo `0b` o `0B`: `0b11000011`

In [6]:
print(123 + 456 - 789)

print(123456789012345678901234567890 + 1)

print(1234567890123456789012345678901234567890 + 1)

resultado = 2 ** 888
print(f"2^888 tiene {len(str(resultado))} dígitos")

print(type(123))

-210
123456789012345678901234567891
1234567890123456789012345678901234567891
2^888 tiene 268 dígitos
<class 'int'>


In [7]:
print(0x1abc)
print(0o1776)
print(0b11000011)

print(hex(1234))
print(bin(254))
print(oct(1234))

6844
1022
195
0x4d2
0b11111110
0o2322


### 6.2 `float` — Números en Punto Flotante

Tipo de dato (`float`) para representar números reales, para uso en aplicaciones científicas o comerciales (Corcuera, IntroPython).

**Características clave:**
- Tienen una representación **IEEE 754 de 64 bits** (típicamente equivalente a `double` en C).
- Tienen **15-17 dígitos decimales de precisión**.
- Para obtener el máximo valor usar `sys.float_info.max`.
- Se puede usar notación científica: `4e5` = 4 × 10⁵.
- No son de tamaño ilimitado como los enteros.

> **Advertencia de precisión:** Los floats son inherentemente imprecisos. Por ejemplo, `0.1 * 3` no da exactamente `0.3` sino `0.30000000000000004`. Para cálculos que requieren precisión exacta (como finanzas), se debe usar el módulo `decimal`.

In [2]:
print(1.23 * -4e5)
print(type(1.2))

import math
print(f"math.pi = {math.pi}")

import random
print(f"random.random() = {random.random()}")

print(f"0.1 * 3 = {0.1 * 3}")

import decimal
x = decimal.Decimal('0.1')
print(f"Decimal('0.1') * 3 = {x * 3}")
print(type(x))

# Error de tamaño de flotante
print(f"1e+1000 = {float(1e1000)}") # Esto devuelve inf (infinito) porque el número es demasiado grande para ser representado como un flotante

-492000.0
<class 'float'>
math.pi = 3.141592653589793
random.random() = 0.9195654618272631
0.1 * 3 = 0.30000000000000004
Decimal('0.1') * 3 = 0.3
<class 'decimal.Decimal'>
1e+1000 = inf


### 6.3 `complex` — Números Complejos

Tipo de dato (`complex`) para representar números complejos de la forma `a + bj` (Corcuera, IntroPython).

**Características clave:**
- La parte imaginaria se denota con `j` (no `i` como en matemáticas).
- Se accede a las partes con `.real` e `.imag`.
- Se puede crear con la función `complex(real, imag)`.
- Soportan operaciones aritméticas directamente.

In [9]:
x = 1 + 2j
print(f"x = {x}")
print(f"Parte real: {x.real}")
print(f"Parte imaginaria: {x.imag}")
print(f"Tipo: {type(x)}")

print(f"x * (3 + 4j) = {x * (3 + 4j)}")

z = complex(2, -3)
print(f"z = {z}")

x = (1+2j)
Parte real: 1.0
Parte imaginaria: 2.0
Tipo: <class 'complex'>
x * (3 + 4j) = (-5+10j)
z = (2-3j)


### 6.4 `bool` — Booleanos

Tipo de dato (`bool`) que tiene solo dos valores: `True` y `False` (Corcuera, IntroPython).

**Características clave:**
- `bool` es una **subclase de `int`**: `True` equivale a `1` y `False` equivale a `0`.
- El entero `0`, un valor vacío (cadena vacía `''`, lista vacía `[]`, tupla vacía `()`, diccionario vacío `{}`), y `None` son tratados como `False`; todo lo demás es `True`.
- Los booleanos se comportan como enteros en operaciones aritméticas.

> **Concepto de "Truthiness" (MoureDev, Guía Junior 2026):** Los booleanos son la base de toda lógica en programación. En Python, cualquier valor puede evaluarse como verdadero o falso en un contexto booleano.

In [ ]:
print(8 == 8)
print(8 == 9)
print(type(True))

print(f"bool(0) = {bool(0)}")
print(f"bool(1) = {bool(1)}")
print(f"bool('') = {bool('')}")
print(f"bool('hola') = {bool('hola')}")
print(f"bool([]) = {bool([])}")
print(f"bool([1,2]) = {bool([1,2])}")
print(f"bool(None) = {bool(None)}")

print(f"True + 3 = {True + 3}")
print(f"False + 1 = {False + 1}")
print(f"isinstance(True, int) = {isinstance(True, int)}") # isinstance() te dice si un objeto es una instancia de una clase o tipo específico, en este caso, verifica si True es considerado un entero (int) en Python.

True
False
<class 'bool'>
bool(0) = False
bool(1) = True
bool('') = False
bool('hola') = True
bool([]) = False
bool([1,2]) = True
bool(None) = False
True + 3 = 4
False + 1 = 1
isinstance(True, int) = True


### 6.5 `NoneType` — El valor `None`

Python proporciona un valor especial llamado `None` que representa la **ausencia intencional de valor** (Corcuera, IntroPython).

**Características clave:**
- Equivalente conceptual al puntero `NULL` de C.
- `None` es el único valor del tipo `NoneType`.
- Se usa `is` y `is not` para comprobar si una variable es `None` (no `==`).
- Es común usarlo para inicializar variables o como valor de retorno por defecto de funciones.

In [11]:
x = None
print(f"x = {x}")
print(f"type(x) = {type(x)}")

print(f"x is None: {x is None}")
print(f"x is not None: {x is not None}")

x = None
type(x) = <class 'NoneType'>
x is None: True
x is not None: False


---
## 7. Conversión de Tipos

Se pueden convertir tipos mediante las funciones integradas `int()`, `float()`, `str()`, `bool()`, etc. (Corcuera, IntroPython).

| Función | Convierte a | Ejemplo |
|---------|-------------|--------|
| `int(x)` | Entero | `int('123')` → `123` |
| `float(x)` | Flotante | `float('3.14')` → `3.14` |
| `str(x)` | Cadena | `str(123)` → `'123'` |
| `bool(x)` | Booleano | `bool(0)` → `False` |

**Conversión implícita:** Python convierte automáticamente enteros y flotantes convenientemente. Por ejemplo, `1 + 2.0` produce `3.0` (el `int` se promueve a `float`).

In [12]:
x = '123'
print(f"Original: {x!r}, tipo: {type(x).__name__}")

x = int(x)
print(f"int(x): {x!r}, tipo: {type(x).__name__}")

x = float(x)
print(f"float(x): {x!r}, tipo: {type(x).__name__}")

x = str(x)
print(f"str(x): {x!r}, tipo: {type(x).__name__}")

x = bool(x)
print(f"bool(x): {x!r}, tipo: {type(x).__name__}")

x = str(x)
print(f"str(x): {x!r}, tipo: {type(x).__name__}")

Original: '123', tipo: str
int(x): 123, tipo: int
float(x): 123.0, tipo: float
str(x): '123.0', tipo: str
bool(x): True, tipo: bool
str(x): 'True', tipo: str


In [ ]:
# Conversión automática en expresiones mixtas, son deterministas así que no es tipado débil.
print(f"1 + 2.0 = {1 + 2.0}")
print(f"type(1 + 2.0) = {type(1 + 2.0)}")

print(f"True + 5 = {True + 5}")
print(f"False * 10 = {False * 10}")

1 + 2.0 = 3.0
type(1 + 2.0) = <class 'float'>
True + 5 = 6
False * 10 = 0


---
## 8. Operadores

Según la Guía de MoureDev (2026), los operadores son los siguientes:
- **Aritméticos:** `+`, `-`, `*`, `/`, `%` (módulo), `**` (potencia)
- **Comparación:** `==`, `!=`, `<`, `>`, `<=`, `>=`
- **Lógicos:** `and`, `or`, `not` (evaluación de cortocircuito)
- **Asignación:** `=`, `+=`, `-=`, `*=`, `/=` (asignación compuesta)
- **Bit a bit:** `&`, `|`, `^`, `~`, `<<`, `>>` (nivel de bits)

### 8.1 Operadores Aritméticos

Python tiene 7 operadores aritméticos (Bonaretti, Learn Python with Jupyter):

| Operador | Descripción | Ejemplo | Resultado |
|----------|-------------|---------|----------|
| `+` | Suma | `4 + 3` | `7` |
| `-` | Resta | `6 - 2` | `4` |
| `*` | Multiplicación | `6 * 5` | `30` |
| `/` | División flotante | `1 / 2` | `0.5` |
| `//` | División entera (floor) | `7 // 4` | `1` |
| `**` | Exponenciación | `2 ** 5` | `32` |
| `%` | Módulo (residuo) | `9 % 2` | `1` |

**Detalles importantes sobre la división (Corcuera, IntroPython):**
- `/` siempre retorna un `float`: `1 / 2` → `0.5`, `-1 / 2` → `-0.5`
- `//` retorna el piso (floor integer): `1 // 2` → `0`, `-1 // 2` → `-1`
- `//` con floats: `8.9 // 2.5` → `3.0`, `-8.9 // 2.5` → `-4.0`
- `%` con negativos: `-9 % 2` → `1`, `9 % -2` → `-1`

In [14]:
print("--- División flotante ---")
print(f"1 / 2 = {1 / 2}")
print(f"-1 / 2 = {-1 / 2}")
print(f"10 / 5 = {10 / 5}")

print("\n--- División entera (floor) ---")
print(f"1 // 2 = {1 // 2}")
print(f"-1 // 2 = {-1 // 2}")
print(f"7 // 4 = {7 // 4}")
print(f"8.9 // 2.5 = {8.9 // 2.5}")
print(f"-8.9 // 2.5 = {-8.9 // 2.5}")

print("\n--- Módulo ---")
print(f"9 % 2 = {9 % 2}")
print(f"-9 % 2 = {-9 % 2}")
print(f"9 % -2 = {9 % -2}")
print(f"7.5 % 4 = {7.5 % 4}")

print("\n--- Exponenciación ---")
print(f"2 ** 5 = {2 ** 5}")
print(f"1.2 ** 3.4 = {1.2 ** 3.4}")

--- División flotante ---
1 / 2 = 0.5
-1 / 2 = -0.5
10 / 5 = 2.0

--- División entera (floor) ---
1 // 2 = 0
-1 // 2 = -1
7 // 4 = 1
8.9 // 2.5 = 3.0
-8.9 // 2.5 = -4.0

--- Módulo ---
9 % 2 = 1
-9 % 2 = 1
9 % -2 = -1
7.5 % 4 = 3.5

--- Exponenciación ---
2 ** 5 = 32
1.2 ** 3.4 = 1.858729691979481


### Funciones integradas para números

Python contiene funciones integradas para manipular números (Corcuera, IntroPython):
- **Matemáticas:** `round()`, `pow()`, `abs()`
- **Conversión de tipos:** `int()`, `float()`, `str()`, `bool()`, `type()`
- **Conversión de base:** `hex()`, `bin()`, `oct()`

In [15]:
x = 1.23456

print(f"round({x}) = {round(x)}")
print(f"round({x}, 1) = {round(x, 1)}")
print(f"round({x}, 2) = {round(x, 2)}")

print(f"pow(2, 5) = {pow(2, 5)}")
print(f"abs(-4.1) = {abs(-4.1)}")

print(f"hex(1234) = {hex(1234)}")
print(f"bin(254) = {bin(254)}")
print(f"oct(1234) = {oct(1234)}")
print(f"0xABCD = {0xABCD}")

round(1.23456) = 1
round(1.23456, 1) = 1.2
round(1.23456, 2) = 1.23
pow(2, 5) = 32
abs(-4.1) = 4.1
hex(1234) = 0x4d2
bin(254) = 0b11111110
oct(1234) = 0o2322
0xABCD = 43981


### 8.2 Operadores de Comparación

Los operadores de comparación se aplican a enteros y flotantes y producen un resultado booleano (Corcuera, IntroPython):

| Operador | Descripción | Ejemplo |
|----------|-------------|--------|
| `<` | Menor que | `2 < 13` → `True` |
| `<=` | Menor o igual que | `2 <= 2` → `True` |
| `>` | Mayor que | `13 > 2` → `True` |
| `>=` | Mayor o igual que | `3 >= 3` → `True` |
| `==` | Igual a | `2 == 3` → `False` |
| `!=` | Diferente de | `3 != 2` → `True` |
| `in` | Contenido en secuencia | `4 in [1,4,3]` → `True` |
| `not in` | No contenido en secuencia | `6 not in [1,4,3]` → `True` |
| `is` | Mismo objeto en memoria | `type(x) is int` |
| `is not` | Diferente objeto en memoria | `type(x) is not int` |

> **Puente con C:** En C, la comparación de igualdad usa `==` igual que Python, pero `is` no tiene equivalente directo. En Python `==` compara **valores** mientras que `is` compara **identidad** (si apuntan al mismo objeto en memoria).

In [16]:
number = 5

print(f"number > 3 = {number > 3}")
print(f"number < 3 = {number < 3}")
print(f"number == 5 = {number == 5}")
print(f"number != 5 = {number != 5}")

result = number > 3
print(f"result = {result}, type = {type(result)}")

lis = [1, 4, 3, 2, 5]
print(f"4 in lis = {4 in lis}")
print(f"6 not in lis = {6 not in lis}")

x = 5
print(f"type(x) is int = {type(x) is int}")

number > 3 = True
number < 3 = False
number == 5 = True
number != 5 = False
result = True, type = <class 'bool'>
4 in lis = True
6 not in lis = True
type(x) is int = True


### 8.3 Operadores Lógicos

A diferencia de C (`&&`, `||`, `!`), Python usa palabras legibles (Corcuera, IntroPython):

| Operador | Descripción | Equivalente en C |
|----------|-------------|-----------------|
| `and` | AND lógico | `&&` |
| `or` | OR lógico | `\|\|` |
| `not` | NOT lógico | `!` |

**Evaluación de cortocircuito (MoureDev, Guía Junior 2026):**
- `and`: si el primer operando es `False`, no evalúa el segundo.
- `or`: si el primer operando es `True`, no evalúa el segundo.

In [17]:
number = 3

print("--- Tabla AND ---")
print(f"True and True = {True and True}")
print(f"True and False = {True and False}")
print(f"False and True = {False and True}")
print(f"False and False = {False and False}")

print("\n--- Combinando condiciones ---")
print(f"number > 1 = {number > 1}")
print(f"number < 5 = {number < 5}")
print(f"number > 1 and number < 5 = {number > 1 and number < 5}")

print(f"\nTrue and not False = {True and not False}")

--- Tabla AND ---
True and True = True
True and False = False
False and True = False
False and False = False

--- Combinando condiciones ---
number > 1 = True
number < 5 = True
number > 1 and number < 5 = True

True and not False = True


### 8.4 Operadores de Bits (Bitwise)

Permiten operaciones a nivel de bits. Funcionan igual que en C (Corcuera, IntroPython):

| Operador | Descripción | Ejemplo |
|----------|-------------|--------|
| `&` | AND bit a bit | `0b10000001 & 0b10001111` → `0b10000001` |
| `\|` | OR bit a bit | `0b10000001 \| 0b10001111` → `0b10001111` |
| `~` | NOT bit a bit | `~0b10000001` → `-0b10000010` |
| `^` | XOR bit a bit | `0b10000001 ^ 0b10001111` → `0b00001110` |
| `<<` | Desplazamiento izquierda | `0b10000001 << 2` → `0b1000000100` |
| `>>` | Desplazamiento derecha | `0b10000001 >> 2` → `0b100000` |

**Nota:**
El operador `~` (tilde) no realiza una simple inversión estética de los dígitos visibles en la pantalla, sino que se define rigurosamente mediante la siguiente relación:

$$\sim x = -x - 1$$

Esto debido a superar las dificultades de representar los números negativos junto al cero.
Si se toma un byte común (8 bits) para evitar el infinito de Python, el número más alto con el bit de signo en cero (positivo) es el 127 (01111111). El número 0 se escribe 00000000.Si aplicamos ~ al cero (00000000), el sistema invierte todos los bits y obtenemos 11111111.En la convención de complemento a dos, la secuencia 11111111 no representa el número más alto, sino el $-1$.De este modo, se cumple la fórmula:$$\sim 0 = -0 - 1 = -1$$El desplazamiento de una unidad ocurre porque el sistema debe dar cabida al cero dentro de la escala numérica sin duplicar su signo —un inconveniente crónico de sistemas más primitivos como el signo-magnitud, donde existían el $+0$ y el $-0$—, pero forzando a que la inversión pura de bits se sitúe siempre un paso por detrás del valor negativo directo.

In [7]:
x = 0b10000001
y = 0b10001111
z = 0b00000000

print(f"x = {bin(x)} ({x})")
print(f"y = {bin(y)} ({y})")
print(f"z = {bin(z)} ({z})")
print(f"x & y  = {bin(x & y)} ({x & y})")
print(f"x | y  = {bin(x | y)} ({x | y})")
print(f"~x     = {bin(~x)} ({~x})") 
print(f"~y     = {bin(~y)} ({~y})")
print(f"~z     = {bin(~z)} ({~z})")
print(f"x ^ y  = {bin(x ^ y)} ({x ^ y})")
print(f"x << 2 = {bin(x << 2)} ({x << 2})")
print(f"x >> 2 = {bin(x >> 2)} ({x >> 2})")

print(f"\n--- Ejemplo simple: 5 & 3 ---")
print(f"5 = {bin(5)} (0101)")
print(f"3 = {bin(3)} (0011)")
print(f"5 & 3 = {5 & 3} ({bin(5 & 3)}) = 0001")

x = 0b10000001 (129)
y = 0b10001111 (143)
z = 0b0 (0)
x & y  = 0b10000001 (129)
x | y  = 0b10001111 (143)
~x     = -0b10000010 (-130)
~y     = -0b10010000 (-144)
~z     = -0b1 (-1)
x ^ y  = 0b1110 (14)
x << 2 = 0b1000000100 (516)
x >> 2 = 0b100000 (32)

--- Ejemplo simple: 5 & 3 ---
5 = 0b101 (0101)
3 = 0b11 (0011)
5 & 3 = 1 (0b1) = 0001


### 8.5 Operadores de Asignación Compuesta

Permiten combinar una operación con la asignación en un solo paso (Corcuera, IntroPython):

| Operador | Ejemplo | Equivalente a |
|----------|---------|---------------|
| `=` | `x = 5` | `x = 5` |
| `+=` | `x += 5` | `x = x + 5` |
| `-=` | `x -= 5` | `x = x - 5` |
| `*=` | `x *= 5` | `x = x * 5` |
| `/=` | `x /= 5` | `x = x / 5` |
| `%=` | `x %= 5` | `x = x % 5` |
| `//=` | `x //= 5` | `x = x // 5` |
| `**=` | `x **= 5` | `x = x ** 5` |
| `&=` | `x &= 5` | `x = x & 5` |
| `\|=` | `x \|= 5` | `x = x \| 5` |
| `^=` | `x ^= 5` | `x = x ^ 5` |
| `>>=` | `x >>= 5` | `x = x >> 5` |
| `<<=` | `x <<= 5` | `x = x << 5` |

In [19]:
x = 10
print(f"x = {x}")

x += 5
print(f"x += 5  → x = {x}")

x -= 3
print(f"x -= 3  → x = {x}")

x *= 2
print(f"x *= 2  → x = {x}")

x //= 3
print(f"x //= 3 → x = {x}")

x **= 2
print(f"x **= 2 → x = {x}")

x = 10
x += 5  → x = 15
x -= 3  → x = 12
x *= 2  → x = 24
x //= 3 → x = 8
x **= 2 → x = 64


---
## 9. Cadenas de Texto (Strings)

Tipo de dato (`str`) para representar cadenas de caracteres, para uso en procesado de textos (Corcuera, IntroPython).

**Características fundamentales:**
- Se delimitan por `'...'`, `"..."`, `'''...'''`, o `"""..."""`.
- Python 3 usa el conjunto de caracteres **Unicode**.
- Los strings son **inmutables**: su contenido no se puede modificar directamente.
- Para convertir números en strings se usa `str()`.
- Para convertir strings a números se usa `int()` o `float()`.
- Python **no tiene un tipo `char` dedicado**: un carácter es simplemente un string de longitud 1.

> **Puente con C:** En C, las cadenas son arrays de `char` terminados en `\0` y son mutables (`str[0] = 'A'`). En Python, los strings son objetos inmutables. Intentar `nombre[0] = 'E'` genera un `TypeError`.

### 9.1 Creación de Strings y Secuencias de Escape

Para especificar caracteres especiales se usan **secuencias de escape**:

| Secuencia | Significado |
|-----------|-------------|
| `\n` | Salto de línea |
| `\t` | Tabulación |
| `\r` | Retorno de carro |
| `\\` | Barra invertida |
| `\'` | Comilla simple |
| `\"` | Comilla doble |

In [20]:
s1 = 'apple'
print(f"s1 = {s1!r}")

s2 = "orange"
print(f"s2 = {s2!r}")

s3 = "'orange'"
print(f"s3 = {s3!r}")

s3 = "\"orange\""
print(f"s3 = {s3!r}")

s4 = """testing
testing"""
print(f"s4 = {s4!r}")

mensaje = "Hola\n\tMundo"
print(mensaje)

s1 = 'apple'
s2 = 'orange'
s3 = "'orange'"
s3 = '"orange"'
s4 = 'testing\ntesting'
Hola
	Mundo


### 9.2 Funciones y Operadores para Strings

| Función/Operador | Descripción | Ejemplo |
|------------------|-------------|--------|
| `len()` | Longitud | `len('Hello')` → `5` |
| `in` | ¿Contenido? | `'ell' in 'Hello'` → `True` |
| `+` | Concatenación | `'Hello' + '!'` → `'Hello!'` |
| `*` | Repetición | `'Hello' * 2` → `'HelloHello'` |
| `[i]`, `[-i]` | Indexación | `'Hello'[1]` → `'e'` |
| `[m:n:step]` | Slicing | `'Hello'[1:3]` → `'el'` |

In [21]:
s = "Hello, world"

print(f"type(s) = {type(s)}")
print(f"len(s) = {len(s)}")
print(f"'ello' in s = {'ello' in s}")

print(f"\n--- Indexación ---")
print(f"s[0] = {s[0]!r}")
print(f"s[1] = {s[1]!r}")
print(f"s[-1] = {s[-1]!r}")
print(f"s[-2] = {s[-2]!r}")

print(f"\n--- Slicing ---")
print(f"s[1:3] = {s[1:3]!r}")
print(f"s[1:-1] = {s[1:-1]!r}")
print(f"s[:4] = {s[:4]!r}")
print(f"s[4:] = {s[4:]!r}")
print(f"s[:] = {s[:]!r}")
print(f"s[0:5:2] = {s[0:5:2]!r}")

print(f"\n--- Concatenación y Repetición ---")
s2 = s + " again"
print(f"s + ' again' = {s2!r}")
print(f"s2 * 2 = {(s2 * 2)!r}")

type(s) = <class 'str'>
len(s) = 12
'ello' in s = True

--- Indexación ---
s[0] = 'H'
s[1] = 'e'
s[-1] = 'd'
s[-2] = 'l'

--- Slicing ---
s[1:3] = 'el'
s[1:-1] = 'ello, worl'
s[:4] = 'Hell'
s[4:] = 'o, world'
s[:] = 'Hello, world'
s[0:5:2] = 'Hlo'

--- Concatenación y Repetición ---
s + ' again' = 'Hello, world again'
s2 * 2 = 'Hello, world againHello, world again'


### 9.3 Inmutabilidad de Strings

Los strings son **inmutables**: no puedes cambiar un carácter individual como harías en un array de caracteres de C (`char str[]`).

In [22]:
nombre = "Ada Lovelace"

# nombre[0] = 'E'  # TypeError: 'str' object does not support item assignment

nombre_modificado = "E" + nombre[1:]
print(f"Original: {nombre}")
print(f"Modificado: {nombre_modificado}")

Original: Ada Lovelace
Modificado: Eda Lovelace


### 9.4 Métodos de Strings

La clase `str` proporciona varios métodos útiles (Corcuera, IntroPython):

| Método | Descripción |
|--------|-------------|
| `s.strip()` | Elimina espacios al inicio y final |
| `s.rstrip()` | Elimina espacios al final |
| `s.lstrip()` | Elimina espacios al inicio |
| `s.upper()` | Convierte a mayúsculas |
| `s.lower()` | Convierte a minúsculas |
| `s.isupper()` | ¿Es todo mayúsculas? |
| `s.islower()` | ¿Es todo minúsculas? |
| `s.find(sub)` | Busca subcadena, retorna índice o -1 |
| `s.index(sub)` | Busca subcadena, retorna índice (error si no existe) |
| `s.startswith(prefix)` | ¿Empieza con...? |
| `s.endswith(suffix)` | ¿Termina con...? |
| `s.split(sep)` | Divide en lista por separador |
| `sep.join(list)` | Une lista de strings con separador |

In [23]:
texto = "  Hola Mundo Python  "

print(f"Original: {texto!r}")
print(f"strip(): {texto.strip()!r}")
print(f"lstrip(): {texto.lstrip()!r}")
print(f"rstrip(): {texto.rstrip()!r}")

s = "Hello, World"
print(f"\nupper(): {s.upper()!r}")
print(f"lower(): {s.lower()!r}")
print(f"isupper(): {s.isupper()}")

print(f"\nfind('World'): {s.find('World')}")
print(f"find('xyz'): {s.find('xyz')}")
print(f"startswith('Hello'): {s.startswith('Hello')}")
print(f"endswith('World'): {s.endswith('World')}")

print(f"\nsplit(', '): {s.split(', ')}")

palabras = ['Python', 'es', 'genial']
print(f"' '.join(palabras): {' '.join(palabras)!r}")

Original: '  Hola Mundo Python  '
strip(): 'Hola Mundo Python'
lstrip(): 'Hola Mundo Python  '
rstrip(): '  Hola Mundo Python'

upper(): 'HELLO, WORLD'
lower(): 'hello, world'
isupper(): False

find('World'): 7
find('xyz'): -1
startswith('Hello'): True
endswith('World'): True

split(', '): ['Hello', 'World']
' '.join(palabras): 'Python es genial'


### 9.5 ¿Tipo carácter? `ord()` y `chr()`

Python no tiene un tipo de dato dedicado a caracteres. Un carácter es un string de longitud 1. Las funciones integradas `ord()` y `chr()` operan sobre strings de longitud 1 (Corcuera, IntroPython):

- `ord(c)`: retorna el ordinal Unicode (entero) del carácter.
- `chr(i)`: retorna el carácter correspondiente al ordinal Unicode.

In [24]:
print(f"ord('A') = {ord('A')}")
print(f"ord('水') = {ord('水')}")
print(f"chr(65) = {chr(65)!r}")
print(f"chr(27700) = {chr(27700)!r}")

ord('A') = 65
ord('水') = 27700
chr(65) = 'A'
chr(27700) = '水'


### 9.6 Formato de Strings

Existen varias formas de formatear strings en Python (Corcuera, IntroPython):

1. **f-strings** (Python 3.6+, la más moderna y recomendada)
2. **`str.format()`** (Python 3.x)
3. **Operador `%`** (estilo C, legacy)

#### f-strings (Format Strings)

La forma más moderna y recomendada. Permiten insertar expresiones directamente dentro del string usando `{}`.

In [25]:
nombre = "Ada Lovelace"
edad = 36
precio = 14.5678

print(f"La programadora {nombre} vivió {edad} años.")
print(f"El precio formateado a 2 decimales es: {precio:.2f}")
print(f"2 + 3 = {2 + 3}")
print(f"{'Hola':>10} | {'Mundo':<10} | {'Python':^10}")

La programadora Ada Lovelace vivió 36 años.
El precio formateado a 2 decimales es: 14.57
2 + 3 = 5
      Hola | Mundo      |   Python  


#### `str.format()`

Usa la función `format()` y `{}` como campos de reemplazo (Corcuera, IntroPython):

In [26]:
print('|{}|{}|more|'.format('Hello', 'world'))

print('|{0}|{1}|more|'.format('Hello', 'world'))
print('|{1}|{0}|more|'.format('Hello', 'world'))

print('|{greeting}|{name}|'.format(greeting='Hello', name='Peter'))

print('|{1:8}|{0:>7}|'.format('Hello', 'Peter'))

print('|{a:.3f}|{b:6.2f}|{c:4d}|'.format(a=1.2, b=3.456, c=78))

|Hello|world|more|
|Hello|world|more|
|world|Hello|more|
|Hello|Peter|
|Peter   |  Hello|
|1.200|  3.46|  78|


#### Alineación con métodos de string

Se pueden usar las funciones `str.rjust(n)`, `str.ljust(n)`, `str.center(n)`, `str.zfill(n)` donde `n` es el ancho de campo (Corcuera, IntroPython):

In [27]:
print(f"'123'.rjust(5) = {'123'.rjust(5)!r}")
print(f"'123'.ljust(5) = {'123'.ljust(5)!r}")
print(f"'123'.center(5) = {'123'.center(5)!r}")
print(f"'123'.zfill(5) = {'123'.zfill(5)!r}")
print(f"'-1.2'.zfill(6) = {'-1.2'.zfill(6)!r}")

'123'.rjust(5) = '  123'
'123'.ljust(5) = '123  '
'123'.center(5) = ' 123 '
'123'.zfill(5) = '00123'
'-1.2'.zfill(6) = '-001.2'


---
## 10. Ejercicio Práctico: Calculadora de Calificaciones Universitarias

**Contexto:** Eres estudiante y necesitas calcular tu promedio ponderado del semestre para 3 materias.

**Instrucciones:**
1. Define variables con los datos de **3 materias** (nombre, nota, y peso/créditos). Usa los siguientes valores de ejemplo:
   - Materia 1: `"Matemáticas"`, nota `8.5`, peso `4`
   - Materia 2: `"Física"`, nota `7.0`, peso `3`
   - Materia 3: `"Programación"`, nota `9.2`, peso `5`
2. Calcula el promedio ponderado
3. Usa `f-strings` para imprimir un reporte final que muestre cada materia con su nota y el promedio final formateado a 2 decimales.

**Restricciones:**
- Aún no hemos visto bucles (`for`/`while`) ni listas (`list`), por lo que debes crear variables individuales para cada materia (ej. `materia1`, `nota1`, `peso1`, etc.).

**Conceptos que aplicarás:**
- Variables y asignación
- Conversión implícita de tipos (int + float)
- Operadores aritméticos (`*`, `+`, `/`)
- f-strings para formateo de salida

¡Escribe tu código en la celda de abajo!

In [28]:
# Escribe tu solución aquí:
materia1 = 'Matemáticas'
nota1 = 8.5
peso1 = 4

materia2 = 'Física'
nota2 = 7.0
peso2 = 3

materia3 = 'Programación'
nota3 = 9.2
peso3 = 5

# Calcula el promedio ponderado

# Imprime el reporte